File: Problems/No.23_Merge_k_Sorted_Lists/merge_k_sorted_lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        
        # parameters, priority queue with accending order
        PQ = []
        mapping = {}
        ans = cur = None
        
        # start compute
        if (len(lists) == 0):
            return None
        elif (len(lists) == 1):
            return lists[0]
        else:
            
            # push the first values into priority queue
            for i in range(len(lists)):
                if lists[i] is not None:
                    heappush(PQ, (lists[i].val, i))
                    mapping[i] = lists[i]

            # start computing, connect the highest priority in PQ until there is nothing in PQ
            while PQ:
                
                val, index = heappop(PQ)
                
                # check if it is the first node
                if ans is None:
                    ans = cur = mapping[index]
                else:
                    cur.next = mapping[index]
                    cur = cur.next
                    
                # check whether the next pointer is null or not
                if mapping[index].next is not None:
                    mapping[index] = mapping[index].next
                    heappush(PQ, (mapping[index].val, index))

            return ans
        

File: Problems/No.23_Merge_k_Sorted_Lists/merge_k_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        
        // parameters, priority queue with accending order
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> PQ;
        pair<int, int> temp;
        ListNode *ans = NULL, *cur = NULL;
        int top_index;
        
        if (lists.size() == 0) return nullptr;
        else if (lists.size() == 1) return lists[0];
        else {
            
            // push the first values into priority queue
            for (int i=0; i<lists.size(); i++) {
                
                if (lists[i] != NULL) {
                    
                    temp.first = lists[i]->val;
                    temp.second = i;
                    PQ.push(temp);
                    
                }
                
            }
            
            // start computing, connect the highest priority in PQ until there is nothing in PQ
            while (PQ.size()) {
                
                top_index = PQ.top().second;
                PQ.pop();
                
                // check if it is the first node
                if (ans == NULL) {
                    
                    ans = lists[top_index];
                    cur = ans;
                    
                }
                else {
                    
                    cur->next = lists[top_index];
                    cur = cur->next;
                    
                }
                
                // check whether the next pointer is null or not
                if (lists[top_index]->next) {
                        
                    lists[top_index] = lists[top_index]->next;
                    temp.first = lists[top_index]->val;
                    temp.second = top_index;
                    PQ.push(temp);

                }
                
            }
            
            return ans;
            
        }
        
    }
};
